pandas/core/arrays/datetimes.py

```diff
@@ -1408,6 +1412,154 @@ def to_julian_date(self):
 DatetimeArrayMixin._add_datetimelike_methods()
 
 
+# -------------------------------------------------------------------
+# Constructor Helpers
+
+def maybe_infer_tz(tz, inferred_tz):
+    """
+    If a timezone is inferred from data, check that it is compatible with
+    the user-provided timezone, if any.
+
+    Parameters
+    ----------
+    tz : tzinfo or None
+    inferred_tz : tzinfo or None
+
+    Returns
+    -------
+    tz : tzinfo or None
+
+    Raises
+    ------
+    TypeError : if both timezones are present but do not match
+    """
+    if tz is None:
+        tz = inferred_tz
+    elif inferred_tz is None:
+        pass
+    elif not timezones.tz_compare(tz, inferred_tz):
+        raise TypeError('data is already tz-aware {inferred_tz}, unable to '
+                        'set specified tz: {tz}'
+                        .format(inferred_tz=inferred_tz, tz=tz))
+    return tz
+
+
+def dtype_conversions(data, copy, has_format=False):
+    """
+    Convert data based on dtype conventions, issuing deprecation warnings
+    or errors where appropriate.
+
+    Parameters
+    ----------
+    data : np.ndarray or pd.Index
+    copy : bool
+    has_format : bool, default False
+        Indicates if the data will be passed to a parsing function with a
+        `format` kwarg.
+
+    Returns
+    -------
+    data : np.ndarray or pd.Index
+    copy : bool
+
+    Raises
+    ------
+    TypeError : PeriodDType data is passed
+    """
+
+    if is_float_dtype(data) and not has_format:
+        # Note: we must cast to datetime64[ns] here in order to treat these
+        #  as wall-times instead of UTC timestamps.
+        data = data.astype(_NS_DTYPE)
+        copy = False
+        # TODO: Why do we treat this differently from integer dtypes?
```

Comment:
`DatetimeIndex(data=int64data, tz=tz)` treats the data as UTC times.  `DatetimeIndex(data=floatdata, tz=tz)` treats the data as wall-times.  I would have expected that float data would be treated like int64 data.

Reply:
yeah i agree this seems a little odd, these should just be cast to integer i think.

Reply:
@jreback i think all comments except this one have been addressed (a few minutes away from green if all goes well).  Changing this would be a breaking change, albeit a small one.  Should we change now, deprecate now, or leave for a later PR?

Reply:
ok for this pr 

yeah should fix / deprecate whatever is needed later